lib/thrift/generator/client.ex

```diff
@@ -0,0 +1,269 @@
+          end
+          defp read_t_application_exception(
+                <<8::size(8),
+                2::16-unsigned,
+                type::32-signed,
+                rest::binary>>, accum) do
+            # read the type
+            exception_type = TApplicationException.exception_type(type)
+            read_t_application_exception(rest, Map.put(accum, :type, exception_type))
+          end
+          defp read_t_application_exception(<<0>>, accum) do
+            # read the field stop and return
+            accum
+          end
+        end
+      end
+    end
+
+    def generate_function(service_module, function) do
+
+      args_module = service_module
+      |> Module.concat(Service.service_module_name(function, :args))
+
+      response_module = Service.service_module_name(function, :response)
+
+      underscored_name = function.name
+      |> Atom.to_string
+      |> Macro.underscore
+      |> String.to_atom
+
+      bang_name = :"#{underscored_name}!"
+
+      vars = function.params
+      |> Enum.map(&Macro.var(&1.name, nil))
+
+      assignments = function.params
+      |> Enum.zip(vars)
+      |> Enum.map(fn {param, var} ->
+        quote do
+          {unquote(param.name), unquote(var)}
+        end
+      end)
+
+      rpc_name = Atom.to_string(function.name)
+
+      quote do
+        def unquote(underscored_name)(client, unquote_splicing(vars), timeout \\ 5000) do
+          serialized_args = %unquote(args_module){unquote_splicing(assignments)}
+          |> unquote(args_module).BinaryProtocol.serialize
+
+          sequence_id = :erlang.unique_integer([:positive])
+          message = Binary.serialize(
+            :message_begin,
+            {unquote(message_type(function)), sequence_id, unquote(rpc_name)})
+
+          GenServer.call(client, {:send, [message | serialized_args]})
+
+          unquote(build_response_handler(function, rpc_name, response_module))
+        end
+        def unquote(bang_name)(client, unquote_splicing(vars), timeout \\ 5000) do
+          case unquote(underscored_name)(client, unquote_splicing(vars), timeout) do
+            {:ok, rsp} ->
+              rsp
+
+            {:error, {:exception, ex}} ->
+              raise ex
+
+            {:error, _}=err ->
+              raise err
+          end
+        end
+      end
+    end
+
+    defp message_type(%Function{oneway: true}), do: :oneway
+    defp message_type(%Function{oneway: false}), do: :call
+
+    defp build_response_handler(%Function{oneway: true}, _, _) do
+      quote do
+        _timeout = timeout
+        {:ok, nil}
+      end
+    end
+    defp build_response_handler(%Function{oneway: false}, rpc_name, response_module) do
+      quote do
+        case GenServer.call(client, {:recv, 0, timeout}) do
+          {:ok, message} ->
+
+            message
+            |> IO.iodata_to_binary
+            |> deserialize_message_reply(unquote(rpc_name), sequence_id, unquote(response_module))
```

Comment:
The ids can become out of sync because of a bug in the client code too. If there is a bug then we want to recover to a good state. It is not clear that a crash will do this.

For example its possible the send GenServer call to timeout but the gen tcp send to succeed. Depending how the connection process is used it would be possible for the next caller to get the result for the previous request when it receives.